### PR TITLE
NETOBSERV-987: Add must gather script for network observability

### PR DESCRIPTION
**Pull request by msherif1234**
New script to collect netobservability logs and enable it by default
```shell
oc adm must-gather
```

Testing with cluster-bot
========================
1- Create must gather for network-observability
```shell
oc adm must-gather
```
2- Check the pods using `omg` tool
```shell
$ omg use must-gather.local.2028201061218736439/
Using:  /home/mmahmoud/data/test-noo-mg/must-gather.local.2028201061218736439/registry-build02-ci-openshift-org-ci-ln-tp5qw22-stable-sha256-da661034dc54b9943a1ce6c9b2abbb56c4ade39d1640ffe13a2f2a8e0867f7e0

$ omg get pods -n netobserv
NAME                               READY  STATUS   RESTARTS  AGE
flowlogs-pipeline-4f8t7            1/1    Running  0         6m53s
flowlogs-pipeline-7slt6            1/1    Running  0         6m53s
flowlogs-pipeline-dbgs4            1/1    Running  0         6m53s
flowlogs-pipeline-ncrzj            1/1    Running  0         6m53s
flowlogs-pipeline-p6n68            1/1    Running  0         6m53s
flowlogs-pipeline-qr554            1/1    Running  0         6m53s
netobserv-plugin-6f448c67d4-c4q4h  1/1    Running  0         6m54s
$ omg get pods -n netobserv-privileged
NAME                        READY  STATUS   RESTARTS  AGE
netobserv-ebpf-agent-2cljh  1/1    Running  0         6m55s
netobserv-ebpf-agent-2fm9j  1/1    Running  0         6m55s
netobserv-ebpf-agent-dsmgb  1/1    Running  0         6m55s
netobserv-ebpf-agent-vgstj  1/1    Running  0         6m55s
netobserv-ebpf-agent-vm7sq  1/1    Running  0         6m55s
netobserv-ebpf-agent-znr24  1/1    Running  0         6m55s
[mmahmoud@mmahmoud test-noo-mg]$ omg get pods -n openshift-netobserv-operator
NAME                                           READY  STATUS   RESTARTS  AGE
netobserv-controller-manager-566fb67dff-2blh5  2/2    Running  0         7m28s
```
3- Checking the flow collector object
```shell
cat cluster-scoped-resources/flows.netobserv.io/flowcollectors/cluster.yaml | head -n 10
apiVersion: flows.netobserv.io/v1beta1
kind: FlowCollector
metadata:
  creationTimestamp: "2023-04-13T16:47:35Z"
  finalizers:
  - flows.netobserv.io/finalizer
  generation: 2
  managedFields:
  - apiVersion: flows.netobserv.io/v1beta1
    fieldsType: FieldsV1
```

### CODE REVIEW

File: collection-scripts/gather_network_observability [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+function get_netobserv_cluster_crs() {
+    declare -a NETOBSERV_CRDS=("flowcollectors")
+    
+    for CRD in "${NETOBSERV_CRDS[@]}"; do
+        oc adm inspect --dest-dir "${BASE_COLLECTION_PATH}" "${CRD}"
+    done
+}
+
+oc adm inspect --dest-dir "${BASE_COLLECTION_PATH}" "ns/${NO_NS}"
+oc adm inspect --dest-dir "${BASE_COLLECTION_PATH}" "ns/netobserv"
```
Comment (jotak) [inline collection-scripts/gather_network_observability:19]: "netobserv" namespace is configurable in the CRD. Can we extract that info from FlowCollector (`spec.namespace`) and inject it in this command ?

**Review by jotak — APPROVED** [review]
tested, lgtm
thanks @msherif1234 !

File: collection-scripts/gather_network_observability [hunk 2]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
+BASE_COLLECTION_PATH="must-gather"
+NO_NS="$(oc get subs -A -o template --template '{{range .items}}{{if eq .spec.name "netobserv-operator"}}{{.metadata.namespace}}{{end}}{{end}}')"
+
+if [ -z "${NO_NS}" ]; then
```
Comment (sferich888) [inline collection-scripts/gather_network_observability:5]: This to me (while perfectly fine); makes me wonder if this should be part of the main must-gather collection; or if this whole collection should be part of a NetObsever collector image. 

Last I checked the NetObserver Operator is not part of the core payload; thus it has its own life cycle and release cadence; meaning that if bugs are found; that team may want to fix them (out of band of the OpenShift release cadance and cycle).  Meaning that this approach to merge the collection of thing into the main collector isn't the right approach.

**Comment by davemulford** [comment]
Since this script is being proposed to be added to the default collection of gathered items, what does this add to the archive in terms of size? What's the size on a 20+ cluster node?

File: collection-scripts/gather_network_observability [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    echo "INFO: Network Observability namespace is not detected. Skipping."
+    exit 0
+fi
+
+function get_netobserv_cluster_crs() {
+    CRD="flowcollectors.flows.netobserv.io"
+    oc adm inspect --dest-dir "${BASE_COLLECTION_PATH}" "${CRD}"
+    obj=$(oc get ${CRD} -o custom-columns=:metadata.name --no-headers)
+    FC_NS=$(oc get ${CRD} ${obj} --output=jsonpath='{.spec.namespace}')
```
Comment (sferich888) [inline collection-scripts/gather_network_observability:14]: The scoping here is odd (I know this works in bash) but it makes code like this very hard to maintain and debug.

File: collection-scripts/gather_network_observability [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+oc adm inspect --dest-dir "${BASE_COLLECTION_PATH}" "ns/${NO_NS}"
+
+get_netobserv_cluster_crs
```
Comment (sferich888) [inline collection-scripts/gather_network_observability:19]: Why are we doing this vs just calling `oc admin inspect` on the cluster resource and letting [inspect](https://github.com/openshift/oc/blob/master/pkg/cli/admin/inspect/resource.go#L51) resolve the resources for the operator (including any CRD's that it defines)? 

@deads2k we may not be able to do this for OLM operators; but if that's true what can we do to fix that (we may have had this conversation before).

**Review by sferich888 — CHANGES_REQUESTED** [review]
I think I would like to see the NetObserver team follow the outline set by the OpenShift logging team in that they create an independant image (or bundle the must-gather collection) with their operator image. 

https://github.com/openshift/cluster-logging-operator/tree/master/must-gather

This has several benefits: 

1. You set the pace of change that occurs with the operator and how it is debugged outside of the OCP release cadence.
2. By nesting the collection with the operator; when  'disconnected' your collection image is already mirrored as part of installation of the operator.

**Review by RickJWagner — APPROVED** [review]
Reads well.  I like @sferich888 comments about using a dedicated image for this purpose.  (See notes at https://github.com/openshift/enhancements/blob/master/enhancements/oc/must-gather.md#must-gather-images).
Otherwise LGTM.

**Pull request merged**